network/p2p/network.go

```diff
@@ -0,0 +1,241 @@
+	defer n.router.lock.Unlock()
+
+	if _, ok := n.router.handlers[handlerID]; ok {
+		return nil, fmt.Errorf("failed to register handler id %d: %w", handlerID, ErrExistingAppProtocol)
+	}
+
+	appRequestTime, err := metric.NewAverager(
+		n.namespace,
+		fmt.Sprintf("handler_%d_app_request", handlerID),
+		"app request time (ns)",
+		n.metrics,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to register app request metric for handler_%d: %w", handlerID, err)
+	}
+
+	appRequestFailedTime, err := metric.NewAverager(
+		n.namespace,
+		fmt.Sprintf("handler_%d_app_request_failed", handlerID),
+		"app request failed time (ns)",
+		n.metrics,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to register app request failed metric for handler_%d: %w", handlerID, err)
+	}
+
+	appResponseTime, err := metric.NewAverager(
+		n.namespace,
+		fmt.Sprintf("handler_%d_app_response", handlerID),
+		"app response time (ns)",
+		n.metrics,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to register app response metric for handler_%d: %w", handlerID, err)
+	}
+
+	appGossipTime, err := metric.NewAverager(
+		n.namespace,
+		fmt.Sprintf("handler_%d_app_gossip", handlerID),
+		"app gossip time (ns)",
+		n.metrics,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to register app gossip metric for handler_%d: %w", handlerID, err)
+	}
+
+	crossChainAppRequestTime, err := metric.NewAverager(
+		n.namespace,
+		fmt.Sprintf("handler_%d_cross_chain_app_request", handlerID),
+		"cross chain app request time (ns)",
+		n.metrics,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to register cross-chain app request metric for handler_%d: %w", handlerID, err)
+	}
+
+	crossChainAppRequestFailedTime, err := metric.NewAverager(
+		n.namespace,
+		fmt.Sprintf("handler_%d_cross_chain_app_request_failed", handlerID),
+		"app request failed time (ns)",
+		n.metrics,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to register cross-chain app request failed metric for handler_%d: %w", handlerID, err)
+	}
+
+	crossChainAppResponseTime, err := metric.NewAverager(
+		n.namespace,
+		fmt.Sprintf("handler_%d_cross_chain_app_response", handlerID),
+		"cross chain app response time (ns)",
+		n.metrics,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to register cross-chain app response metric for handler_%d: %w", handlerID, err)
+	}
+
+	n.router.handlers[handlerID] = &meteredHandler{
+		responder: &responder{
+			Handler:   handler,
+			handlerID: handlerID,
+			log:       n.log,
+			sender:    n.sender,
+		},
+		metrics: &metrics{
+			appRequestTime:                 appRequestTime,
+			appRequestFailedTime:           appRequestFailedTime,
+			appResponseTime:                appResponseTime,
+			appGossipTime:                  appGossipTime,
+			crossChainAppRequestTime:       crossChainAppRequestTime,
+			crossChainAppRequestFailedTime: crossChainAppRequestFailedTime,
+			crossChainAppResponseTime:      crossChainAppResponseTime,
+		},
+	}
```

Comment:
I feel like this chunk should probably be a function on the router - rather than directly modifying the router here (currently we don't use the metrics other than the router... But I could imagine wanting to use them for the client as well, so we could either keep them in the router of move them out 🤷).

The client logic below here should stay here in the network though imo.

We could just leave this as the TODO we already have here - but I feel like doing this would probably reduce the diff and address that TODO.

Reply:
I moved this back into the router where it creates a client with defaults, and then the network is responsible for applying the configured options on top of the default client.